lib/foreman/renderer/scope/macros/base.rb

```diff
@@ -91,64 +177,156 @@ def default_template_url(template, hostgroup)
+              executed while previewing the template output. Examples may be performance
+              heavy operations, destructive operations etc."
+            returns one_of: [true, false], desc: 'A boolean value, true for preview mode, false otherwise'
+            example '<%= preview? ? "# skipping in preview mode" : @host.facts_hash["ssh::rsa::fingerprints::sha256"] -%>'
+          end
           def preview?
             mode == Renderer::PREVIEW_MODE
           end
 
+          apipie :method, 'Generates a random hex string' do
+            required :n, Integer, desc: 'the argument n specifies the length of the random length. The length of the result string is twice of n.'
+            returns String, desc: 'String composed of n random numbers in range of 0-255 in hexadecimal format'
+            example 'rand_hex(5) # => "3bf14f69c1"'
+          end
           def rand_hex(n)
             SecureRandom.hex(n)
           end
 
+          apipie :method, 'Generates a random name' do
+            returns String, desc: 'A random name that can be used as a hostname. The same function is used to suggest
+              a name when provisioning a new host. The format is two names separated by dash.'
+            example 'rand_name # => "addie-debem"'
+          end
           def rand_name
             NameGenerator.new.generate_next_random_name
           end
 
+          apipie :method, 'Generates a text representation of a mac address' do
+            required :mac_address, String, desc: 'mac address in a format of hexadecimal numbers separated by colon'
+            returns String, desc: 'A name that can be used as a hostname. It is based on passed mac-address, same
+              mac-address results in the same hostname.'
+            example 'mac_name("00:11:22:33:44:55") # => "hazel-diana-maruyama-feltus"'
+          end
           def mac_name(mac_address)
             NameGenerator.new.generate_next_mac_name(mac_address)
           end
 
+          apipie :method, 'Returns a kernel release installed on a host based on facts reported by facters' do
+            desc "Given this is based on facts, if it's rendered for multiple hosts in a single template rendering,
+              it's advised to preload `kernel_release` association, see an example below."
+            required :host, 'Host::Managed', desc: 'host object for which kernel released is returned'
+            returns String, desc: 'String describing the kernel release'
+            example 'host_kernel_release(host) # => "3.10.0-957.10.1.el7.x86_6"'
+            example "<%- load_hosts(preload: :kernel_release).each_record do |host| -%>\n<%=   host.name -%>, <%=  host_kernel_release(host) %>\n<%- end -%>"
+          end
           def host_kernel_release(host)
             host&.kernel_release&.value
           end
 
+          apipie :method, 'Returns a host uptime in seconds based on facts reported by facters' do
```

Comment:
ditto, actually you can drop "reported by facters". It's just a fact, it can be reported by Ansible too.